### ISSUE TITLE
Kicker or Ramp stuck ball on VPX 10.8 Standalone (Mac ARM64) 

### ISSUE THREAD

**Issue by tamburro92** [vpinball/vpinball#555]
**Describe the bug**
A specific Kicker or Ramp doesn't work as expected on  table "JP's Lord of the Rings (Stern 2003) v4.vpx". I also noticed a different behavior compared to Windows release, here on mac release instead ball is stuck every time in "Gimli hole" (kicker), it jumps from kicker but again enters in the hole, again and again until forever.
Instead on windows release, kicker shots the ball to the ramp, this is the correct behavior.

Unfortunately I don't know how attach more details or a detailed log.
  
**To Reproduce**
Send ball to Gimli Hole (kicker)

**Expected behavior**
Kicker should shot the ball to the ramp

**Video**
https://github.com/vpinball/vpinball/assets/18575612/6b506190-f32b-490f-a579-eb8fc7d26824

**Crash dump**

**Versions & Settings**
 - VPX [1.8 branch standalone (2c51d41)]
 - Mac ARM64 M1
 - MacOs 13.4 (22F66)
 - stock VPinballX.ini file
 - table JP's Lord of the Rings (Stern 2003) v4
 - rom LOTR 10 en

**Comment by tamburro92** [comment]
Yes, you have to nudge in a perfect timeframe, if you are skilled enough it happens the ball goes straight to ramp and all works good as expected. But It's really hard to nudge in that timeframe, often the nudging has no effect at all, and you need try again in the next time slot.

**Comment by jsm174** [comment]
TBH, I don't really know any of the ball / physics logic. Ie, maybe it's a math / precision error. @toxieainc, any idea of where I could event start to look at?

**Comment by toxieainc** [comment]
@jsm174 You could try disabling the usage of sse2neon, if that changes anything, please.

**Comment by jsm174** [comment]
Thanks @toxieainc. I disabled it, but was able to reproduce this issue. 

I know I definitely had it disabled:

```
[build] #pragma message ("Warning: No SSE kdtree tests")
.
[build] #pragma message ("Warning: No precise divide")
``` 

I don't see this happening with `Lord of the Rings (Stern 2003)_VPW_v1.3_10.6.vpx`.

Maybe it's in the vbs engine?  If we could compare these two tables, might be able to figure it out?


Btw, I've seen something similar in the example table?

https://github.com/vpinball/vpinball/assets/1197137/fe22545a-6540-4cbb-bc9b-3c5ccba3bcca

**Comment by toxieainc** [comment]
This is at least good (the sse2neon part), so what happens in a no-optimization/debug build? The same?

Apart from that, the vbs theory may be good, too.

**Comment by tamburro92** [comment]
@jsm174 I check vbs script of both table, but I can't find any logic related to kicker.
As suggested by @toxieainc could be a bit of difference in engine computations...
Do you think if I can tweak some settings in VPinballX.ini file in order to address these issues or really I have to tweak the table?

**Comment by toxieainc** [comment]
No, there is no such setting, especially as we do not know exactly where/why this happens yet.

**Comment by tamburro92** [comment]
In the next days I'll try to do some compilation changing, I noticed windows compiler [config](https://github.com/vpinball/vpinball/blob/master/VisualPinball.net2022.vcxproj) uses FloatingPointModel: Fast
instead cmake MacArm64 set to default (precise)...
I'll keep you informed.

**Comment by toxieainc** [comment]
Good point, wasn't aware of that. Makes sense in general, as relying on degen floats and other stuff is shaky anyway.

**Comment by tamburro92** [comment]
Good news guys,
It seems with option -ffp-model=strict in makefile bug has disappeared 🔥, no more sutck on LOTR ramp.
I tried before also -ffp-model=fast without success, but we have to do more test about the behavior.
You can look at clang documentation about the implications of different floating point model
https://clang.llvm.org/docs/UsersManual.html#controlling-floating-point-behavior

**Comment by jsm174** [comment]
Awesome, I will give that a try now:

I tried adding this to CMakeLists.txt:

```
add_compile_options(
   -fvisibility=hidden
   -ffast-math
   -Wno-inconsistent-missing-override
   -Wno-missing-exception-spec
   -Wno-shorten-64-to-32
)
```

The ball still never goes up the kicker, but after three or four attempts it will finally get loose and roll down the table.

**Comment by jsm174** [comment]
wow. I added:

```
add_compile_options(
   -fvisibility=hidden
   -ffp-model=strict
   -Wno-inconsistent-missing-override
   -Wno-missing-exception-spec
   -Wno-shorten-64-to-32
)
```

And I can also confirm it works. I will update all the standalone `CMakeLists.txt` files

Great catch!! Thanks!

**Comment by jsm174** [comment]
So looks like this might be a little more involved.

I can't find a universal way to do this via gcc and clang, and then on mac it identifies as AppleClang

```
target_compile_options(vpinball PUBLIC
   $<$<CXX_COMPILER_ID:Clang>:-ffp-model=strict>
   $<$<CXX_COMPILER_ID:AppleClang>:-ffp-model=strict>
   -fvisibility=hidden
   -Wno-inconsistent-missing-override
   -Wno-missing-exception-spec
   -Wno-shorten-64-to-32
)
```

I've tried a bunch of settings for GCC 12, but no luck.

**Comment by toxieainc** [comment]
Could you please retry with
-ffast-math
-ffp-contract=off

or

-fdenormal-fp-math=PreserveSign
-freciprocal-math
-fassociative-math
-fapprox-func
-ffp-contract=fast
(or the last one ffp-contract=off)

Just so that i could get an impression what could be really causing this.

**Comment by jsm174** [comment]
I just tried with 
```
-ffast-math
-ffp-contract=off
```
and 
```
-ffp-contract=off
```

And these both work in `GCC` and `Clang`.

Do you want me to do the other one as well?

**Comment by toxieainc** [comment]
Thats good. So its most likely the increased precision (or some changed instruction order) due to the generated FMAs.
Then i would suggest to add
```
-ffast-math
-ffp-contract=off
```
as the new compilation options instead of `-ffp-model=strict`.

**Comment by toxieainc** [comment]
Oh, maybe one more experiment:
```
-ffast-math
-ffp-contract=on
```
Would be interesting if that also works.

**Comment by jsm174** [comment]
Cool. I just tried with 

```
-ffast-math
-ffp-contract=on
```

And it didn't work. 

I've updated all the standalone cmake files to use 

```
-ffast-math
-ffp-contract=off
```

Closing this issue. Many thanks @tamburro92  and @toxieainc !

**Comment by jsm174** [comment]
Actually opening this back up.

```
   -ffast-math
   -ffp-contract=off
```

is only working with GCC 12, not Clang.. Ughh..

**Comment by tamburro92** [comment]
> Actually opening this back up.
> 
> ```
>    -ffast-math
>    -ffp-contract=off
> ```
> 
> is only working with GCC 12, not Clang.. Ughh..

Confirm, with clang still doesn't work, same bug happens

**Comment by toxieainc** [comment]
Just to double check, if you say GCC12 vs Clang, this is still both macOS, right?
What about
```
-fdenormal-fp-math=PreserveSign
-freciprocal-math
-fassociative-math
-fapprox-func
-ffp-contract=off
```

**Comment by tamburro92** [comment]
```
-fdenormal-fp-math=preserve-sign
-freciprocal-math
-fassociative-math
-fapprox-func
-ffp-contract=off
```
Yes, in this way is working fine with clang on mac arm64 (just note preserve-sign instead of PreserveSign)

**Comment by toxieainc** [comment]
I don't think this will change/improve anything substantially, but you could also try adding to that
```
-fno-math-errno
-fno-trapping-math
```

**Comment by tamburro92** [comment]
I try to summarize the results of the different compilation modes that are working fine ( no stuck ball etc), referring as 1-2-3 on JP's Lord of the Rings (Stern 2003) v4 and clang MacArm64 build.

I noticed that different models change a lot the performance of the table, but the most fluid model is the one with the stuck bug (ffp-model=precise), instead the other modes have worse performance, like more friction, less speed etc.

1st.
`
   -ffp-model=strict
`
2nd.
`
  -fdenormal-fp-math=preserve-sign
  -freciprocal-math
  -fassociative-math
  -fapprox-func
  -ffp-contract=off
`
3rd.
`  -fdenormal-fp-math=preserve-sign
   -freciprocal-math
   -fassociative-math
   -fapprox-func
   -ffp-contract=off
   -fno-math-errno
   -fno-trapping-math
`
I start to describe 2nd mode that is the worst (from my perspective): the ball runs very slow like there is a lot of friction on the table, and when it's hit (by flipper or objects) sometime has a strange changing of trajectory, feeling less natural.
The 1st and 3rd are quite similar and both work better than 2nd, ball has more speed and friction is lesser, maybe the 3rd works a little better...

That is my personal opinion and I think more tests on more tables and together to different floating models should be done before considered closed the issue.
Floating model changes a lot the final performance of the simulator.

**Comment by toxieainc** [comment]
Thanks a lot for your continued experiments. Is this phenomenon with the varying friction specific to this table, or also happening with other (all?) tables.

**Comment by tamburro92** [comment]
@toxieainc Can you suggest other compilation combo modes about float rounding?
It seems 1st. -ffp-model=strict is quite similar to Windows VPX experience, but it still miss something...

**Comment by toxieainc** [comment]
Is this latter one the same as `strict` then?

And i fear going forward we rather should track down which file (and then later on which part of the code in there) is causing these physics differences and then try to properly fix that code so that its more compiler independent.

**Comment by toxieainc** [comment]
And also see my previous question, please.

**Comment by tamburro92** [comment]
The last configuration is similar to strict, but some options are still different.
<img width="548" alt="Screenshot 2023-07-16 alle 12 15 13" src="https://github.com/vpinball/vpinball/assets/18575612/0515fc10-1757-4880-9583-4ad97564ee15">
About your last question, yes floating models affect all physic table.

In my opinion at this time the current source code has already been optimized to work with Visual Studio compiler, furthermore all tables have been adjusted to fit the current build.
So this optimization requires a double effort: 1. adjusting code (for multiple arch) and 2. micro changing to all table configuration

Anyway if I have time i could try to identify the source file impacted ( I think related to physic engine)
